### PR TITLE
refactor(nuxt): enable strict type checking

### PR DESCRIPTION
**Pull request by antfu**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

The first step migrating the entire repo to `strict: true`

- Added temporary `tsconfig.json` to mark the scope of strict - later after we migrate all modules to strict, we could then remove all the sub tsconfig and turn the root level `strict: true`
- `wrapInRef` is redundant - ref already does that: https://github.com/vuejs/core/blob/4d7803ed28fb67d45a83d3500f5407754e65bf64/packages/reactivity/src/ref.ts#L90-L95
- Not only would the strictness help us internally, but also for the generated types consumed by users. Previously the return type of `AsyncData` does not indicate the data could possibly be `null`, this PR also catches that.

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/nuxt/src/app/entry.ts [hunk 1]
```diff
@@ -35,7 +35,7 @@ if (process.server) {
       await nuxt.hooks.callHook('app:created', vueApp)
     } catch (err) {
       await nuxt.callHook('app:error', err)
-      ssrContext.error = ssrContext.error || err
+      ssrContext!.error ||= err
```
Comment (pi0) [inline packages/nuxt/src/app/entry.ts:38]: Node 14 is not supported. Let's delay using this syntax.

File: packages/nuxt/src/app/components/nuxt-link.ts [hunk 2]
```diff
@@ -176,15 +176,21 @@ export function defineNuxtLink (options: NuxtLinkOptions) {
           // converts `""` to `null` to prevent the attribute from being added as empty (`rel=""`)
           : firstNonUndefined<string | null>(props.rel, options.externalRelAttribute, href ? DEFAULT_EXTERNAL_REL_ATTRIBUTE : '') || null
 
-        const navigate = () => navigateTo(href, { replace: props.replace })
+        const navigate = () => {
+          if (href) {
+            navigateTo(href, { replace: props.replace })
```
Comment (pi0) [inline packages/nuxt/src/app/components/nuxt-link.ts:181]: `navigateTo` returns a value (Promise) in some cases. We need to return it.

File: packages/nuxt/src/app/components/nuxt-link.ts [hunk 3]
```diff
@@ -176,15 +176,21 @@ export function defineNuxtLink (options: NuxtLinkOptions) {
...
         if (props.custom) {
-          if (!slots.default) { return null }
+          if (!slots.default) {
+            return null
+          }
           return slots.default({
             href,
             navigate,
-            route: router.resolve(href),
+            route: href ? router.resolve(href) : undefined,
```
Comment (pi0) [inline packages/nuxt/src/app/components/nuxt-link.ts:193]: Can you please double check for behavior of `routerResolve(undefined)`? Does it returns undefined?

Comment (antfu) [reply packages/nuxt/src/app/components/nuxt-link.ts:193]: This part confuses me, it seems `href` could be `null` while those functions do not seem to support them. I will check

Comment (antfu) [reply packages/nuxt/src/app/components/nuxt-link.ts]: image

Just tried, `router.resolve(null)` will result in this error. Probably a potential bug that we never hit. What do you think the best approach for this?

File: packages/nuxt/src/core/app.ts [hunk 4]
```diff
@@ -77,7 +81,8 @@ export async function resolveApp (nuxt: Nuxt, app: NuxtApp) {
 
   // Resolve layouts/ from all config layers
   app.layouts = {}
-  for (const config of nuxt.options._layers.map(layer => layer.config)) {
+  for (const config of getLayerConfigs(nuxt)) {
+    if (!config.srcDir) { continue }
```
Comment (antfu) [inline packages/nuxt/src/core/app.ts:85]: I am not sure about this. Currently, both of these are optional on type:

```ts
nuxt.options._layers.map(layer => layer.config?.srcDir?.startsWith())
//                                       ^       ^
```

While the usage seems to assume them to be not nullable. Is there a missing logic for the runtime, or is there a misalignment in the type?

Comment (pi0) [reply packages/nuxt/src/core/app.ts:85]: You can assume both exist 👍🏼  (we filter layers in nuxi check code)

Comment (pi0) [reply packages/nuxt/src/core/app.ts:85]: I think we should fix the type that this prop exists (btw pleas double check code i'm not 100% sure)

Comment (pi0) [reply packages/nuxt/src/core/app.ts]: Just checked the code. We make sure layers have srcDir: https://github.com/nuxt/framework/blob/cb98c8b9216cb731dc90dcc59195db4d8c244374/packages/kit/src/loader/config.ts#L27

**Review by pi0 — APPROVED** [review]

**Comment by pi0** [comment]
This is a nice initiative 🚀 Merging to avoid stalling PR. There has been some leftovers also in packages/nuxt which we can improve type coverage and fix errors.

I've triple checked but there might be regressions with changes. /cc @danielroe heads up

@antfu Can we please allow unplugin define to make input options mandatory? It can help to properly type them. 🙏🏼

**Comment by antfu** [comment]
@pi0 Published `unplugin@0.9.1` support required options

**Pull request merged**